src/main/java/uk/nhs/digital/nhsconnect/lab/results/inbound/EdifactParser.java

```diff
@@ -0,0 +1,146 @@
+            var messageTrailerIndex = singleMessageEdifactSegments.size() - 1;
+            transactionEndIndexes.add(messageTrailerIndex);
+        }
+
+        var transactionStartEndIndexPairs = zipIndexes(transactionStartIndexes, transactionEndIndexes);
+
+        return transactionStartEndIndexPairs.stream()
+                .map(transactionStartEndIndexPair ->
+                        singleMessageEdifactSegments.subList(transactionStartEndIndexPair.getLeft(), transactionStartEndIndexPair.getRight()))
+                .map(Transaction::new)
+                .collect(Collectors.toList());
+    }
+
+    private List<Pair<Integer, Integer>> zipIndexes(List<Integer> startIndexes, List<Integer> endIndexes) {
+        if (startIndexes.size() != endIndexes.size()) {
+            throw new ToEdifactParsingException(
+                    "Message header-trailer count mismatch: " + startIndexes.size() + "-" + endIndexes.size());
+        }
+
+        var indexPairs = Streams.zip(startIndexes.stream(), endIndexes.stream(), Pair::of)
+                .collect(Collectors.toList());
+
+        if (!areIndexesInOrder(indexPairs)) {
+            throw new ToEdifactParsingException("Message trailer before message header");
+        }
+
+        return indexPairs;
+    }
+
+    private boolean areIndexesInOrder(List<Pair<Integer, Integer>> messageIndexPairs) {
+        // trailer must go after header
+        // next header must go after previous trailer
+        return Comparators.isInOrder(
+                messageIndexPairs.stream()
+                        .map(pair -> List.of(pair.getLeft(), pair.getRight()))
+                        .flatMap(Collection::stream)
+                        .collect(Collectors.toList()),
+                Comparator.naturalOrder());
+    }
+
+    private List<String> extractInterchangeEdifactSegments(List<String> allEdifactSegments) {
+        var firstMessageHeaderIndex = findAllIndexesOfSegment(allEdifactSegments, MessageHeader.KEY).get(0);
+        var allMessageTrailerIndexes = findAllIndexesOfSegment(allEdifactSegments, MessageTrailer.KEY);
+        var lastMessageTrailerIndex = allMessageTrailerIndexes.get(allMessageTrailerIndexes.size() - 1);
+
+        var segmentsBeforeFirstMessageHeader = allEdifactSegments.subList(0, firstMessageHeaderIndex);
+        var segmentsAfterLastMessageTrailer = allEdifactSegments.subList(lastMessageTrailerIndex + 1, allEdifactSegments.size());
+
+        return Stream.of(segmentsBeforeFirstMessageHeader, segmentsAfterLastMessageTrailer)
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+    }
+
+    private List<Integer> findAllIndexesOfSegment(List<String> list, String key) {
+        return IntStream.range(0, list.size())
+                .filter(i -> list.get(i).startsWith(key))
+                .mapToObj(Integer::valueOf)
+                .collect(Collectors.toList());
```

Comment:
IntelliJ should be giving you a warning here: `Can be replaced with 'boxed'`
```suggestion
        return IntStream.range(0, list.size())
                .filter(i -> list.get(i).startsWith(key))
                .boxed()
                .collect(Collectors.toList());
```

Reply:
Change this to `boxed()`